ext/opentelemetry-ext-opentracing-shim/src/opentelemetry/ext/opentracing_shim/__init__.py

```diff
@@ -0,0 +1,289 @@
+        return self._otel_span
+
+    def set_operation_name(self, operation_name):
+        self._otel_span.update_name(operation_name)
+
+        return self
+
+    def finish(self, finish_time=None):
+        end_time = finish_time
+        if end_time is not None:
+            end_time = util.time_seconds_to_ns(finish_time)
+        self._otel_span.end(end_time=end_time)
+
+    def set_tag(self, key, value):
+        self._otel_span.set_attribute(key, value)
+
+        return self
+
+    def log_kv(self, key_values, timestamp=None):
+        if timestamp is not None:
+            event_timestamp = util.time_seconds_to_ns(timestamp)
+        else:
+            event_timestamp = None
+
+        event_name = util.event_name_from_kv(key_values)
+        self._otel_span.add_event(event_name, event_timestamp, key_values)
+
+        return self
+
+    def set_baggage_item(self, key, value):
+        logger.warning(
+            "Calling unimplemented method set_baggage_item() on class %s",
+            self.__class__.__name__,
+        )
+        # TODO: Implement.
+
+    def get_baggage_item(self, key):
+        logger.warning(
+            "Calling unimplemented method get_baggage_item() on class %s",
+            self.__class__.__name__,
+        )
+        # TODO: Implement.
+
+    # TODO: Verify calls to deprecated methods `log_event()` and `log()` on
+    # base class work properly (it's probably fine because both methods call
+    # `log_kv()`).
+
+
+class ScopeWrapper(opentracing.Scope):
+    """A `ScopeWrapper` wraps the OpenTelemetry functionality related to span
+    activation/deactivation while using OpenTracing `Scope` objects for
+    presentation.
+
+    There are two ways to construct a `ScopeWrapper` object: using the `span`
+    argument and using the `span_cm` argument. One and only one of `span` or
+    `span_cm` must be specified.
+
+    When calling the initializer while passing a `SpanWrapper` object in the
+    `span` argument, the `ScopeWrapper` is initialized as a regular `Scope`
+    object.
+
+    When calling the initializer while passing a context manager *generator* in
+    the `span_cm` argument (as returned by the `use_span()` method of
+    OpenTelemetry `Tracer` objects,  the resulting `ScopeWrapper` becomes
+    usable as a context manager (using `with` statements).
+
+    It is necessary to have both ways for constructing `ScopeWrapper`
+    objects because in some cases we need to create the object from a context
+    manager, in which case our only way of retrieving a `Span` object is by
+    calling the `__enter__()` method on the context manager, which makes the
+    span active in the OpenTelemetry tracer; whereas in other cases we need to
+    accept a `SpanWrapper` object and wrap it in a `ScopeWrapper`.
+    """
+
+    def __init__(self, manager, span=None, span_cm=None):
+        super().__init__(manager, span)
+        self._span_cm = span_cm
+
+    def close(self):
+        self._span.unwrap().end()
```

Comment:
You still can't do this, i.e.

```python
with tracer.start_active_span(span, finish_on_end=False):
   # When this block is done, span *will* be finished.
```

Once you decorate an OTel `Span` with its related `SpanWrapper`, you will be able to fetch it and put together the remaining pieces, to this can stay the way it is now, but please add a note here as well.

Reply:
Also, shouldn't `close` and `__exit__` be equivalent? I don't have the OpenTracing implementation open right now but just from intuition that's what I would expect. We should also think about whether we should already call `__enter__` in `__init__`, if that would match the semantics of OpenTracing.

Reply:
@carlosalberto I apologize if I'm missing your point here, but I can't see the problem.

- We accept a boolean `finish_on_close` on `start_active_span()`.
- We pass the value of `finish_on_close` to `ScopeManagerShim.activate()`.
- `ScopeManagerShim.activate()` passes the `finish_on_close` value to the `end_on_exit` argument when calling `use_span()`.
- The `use_span()` implementation on the tracer (the OTel SDK in this case) checks `end_on_exit` and calls `span.end()` only when true.

The tests verify that the span doesn't finish when the scope is exited if `finish_on_close` is false:

https://github.com/kinvolk/opentelemetry-python/blob/6dac24d1b16ee419bb063fbcf8e42afd903feb29/ext/opentelemetry-ext-opentracing-shim/tests/test_shim.py#L169-L176

Could you please tell me how to replicate the problematic state? A code snippet or a failing test could really help here I think.

Reply:
>  shouldn't close and __exit__ be equivalent? I don't have the OpenTracing implementation open right now but just from intuition that's what I would expect.

Correct, this is the case :)

Reply:
@johananl The case is that in OT, calling `__exit__()` and `close()` for `Scope` are equivalent.

Reply:
Thanks for clarifying @carlosalberto. This should (hopefully) be addressed now.